### ISSUE TITLE
File upload problem

### ISSUE THREAD

**Issue by rdvojmoc** [aspnet/KestrelHttpServer#1913]
Problem occurs when validation is done on file and BadRequest is returned from the controller to the client (file to big, or wrong content type), connection is closed after whole file is transferred to the server and not immediately when BadRequest is set in a response  as it should be. 

![snap3677](https://user-images.githubusercontent.com/13537823/27398210-50b96a10-56b9-11e7-92b7-eb5721290b82.png)

Project that i made that reproduce problem:
[FileUpload](https://github.com/rdvojmoc/FileUpload)
There is no part which saves the file to the disk to keep things simple. Problems becomes more obvious when you upload large files ( > 1GB).

Custom model binder (only model state error is set to make things simple):
```
public class AssetBinder : IModelBinder
{
      protected readonly IHostingEnvironment _hosting;
      protected readonly FormFileOptions _fileOptions;	

      public AssetBinder(IHostingEnvironment hosting, IOptions<FormFileOptions> fileOptions)
      {
         _hosting = hosting;
			
         _fileOptions = fileOptions.Value;
      }

      public Task BindModelAsync(ModelBindingContext bindingContext)
      {
          bindingContext.ModelState.TryAddModelError("test", "test error message");
            
          return TaskCache.CompletedTask;
      }
}
```

Controller:
```
[HttpPost]
[DisableFormValueModelBinding]
public IActionResult Post(Asset asset)
{
	if (!ControllerContext.ModelState.IsValid)
	{
	       return BadRequest(ControllerContext.ModelState);
	}
	return Ok();
}
```

I'am testing on Windows 10 Pro, version 1607, build 14393.1358, 64 bit. NET Core version is 1.1 and Microsoft.AspNetCore.Mvc is 1.1.2

**Comment by Tratcher** [comment]
By design. There's nothing inherent about a 400 (Bad Request) status code that says the connection should be closed. HTTP/1.1 requests use KeepAlive connections by default, so if you don't complete reading the request body in your application then Kestrel will attempt to drain it so it can re-use the connection for the next request. The two minute time you see sounds like the idle connection timeout.

KeepAlive can be disabled by setting a "Connection: close" header on the response. This should prevent any draining. @halter73 Kestrel checks for Connection close on response headers right?

**Comment by rdvojmoc** [comment]
Thanks for the insight. I can confirm that setting header "Connection:close" prevents draining.

**Comment by halter73** [comment]
Setting the "Connection: close" header will close the connection only after both the request and the response are completed.

If you want to prevent draining, you can call `HttpContext.Abort()` which will close the connection immediately. Be careful though, because it's possible that the response status line and headers won't be flushed if they weren't flushed already.

**Comment by halter73** [comment]
I think I misinterpreted the question. If you set the "Connection: close" response or response header, the response will still get drained, but I can confirm that in all released versions of Kestrel so far, the request body will *not* be drained if it wasn't already since there isn't much point.

I'm not sure how much you want to rely on this, because I could see potentially attempting to drain the request body despite the "Connection: close" header in the future for validation purposes. Fortunately, starting in 2.0, Kestrel will have a 5 second automatic request body draining timeout.

**Comment by rdvojmoc** [comment]
I tested a bit more and after setting "Connection : close" on the response error 504 gets returned to the client. What can I do if I want to return BadRequest and still have the request drained, before Kestrel 2.0 is released?

**Comment by halter73** [comment]
The 504 originates from IIS not Kestrel. It indicates that the response didn't complete within the configured [requestTimeout](https://docs.microsoft.com/en-us/aspnet/core/hosting/aspnet-core-module#attributes-of-the-aspnetcore-element).

Kestrel will not automatically flush responses to keep-alive requests until the entire request body is drained. The reason for waiting to flush the response is so that Kestrel can return a 400 status for things like bad request chunk encoding even when the app doesn't consume the entire request body.

Setting the "Connection: close" header and ensuring the Task returned by your middleware/controller action/whatever completes should be enough to ensure the response gets autoflushed. If you wan't to be extra sure your response gets flushed before the default 2 minute request timeout, you could also call `context.Response.Body.FlushAsync()` to manually flush the response, but that should be unnecessary.

**Comment by rdvojmoc** [comment]
Kestrel is running without IIS at the moment. Response 504 was shown in Fiddler. If I do same request from Chrome or Firefox I get ERR_CONNECTION_ABORTED and status code 0 after setting "Connection:close" header on the response.  

My goal is to get BadRequest returned right after this is called in my controller:
`if (!ControllerContext.ModelState.IsValid)
{
	return BadRequest(ControllerContext.ModelState);
}`

**Comment by aspnet-hello** [comment]
We periodically close 'discussion' issues that have not been updated in a long period of time.

We apologize if this causes any inconvenience. We ask that if you are still encountering an issue, please log a new issue with updated information and we will investigate.
